src/core/network/client.ts

```diff
@@ -82,6 +82,23 @@ function includesMedia(payload: Record<string, unknown>) {
   })
 }
 
+// Types are pretty wrong if applied to a Partial but
+// is passable if applied to an already defined object
+// E.g. compactOptions({ a: number, b?: number })
+// produces Type { a: number }
+type CompactKeys<T> = Exclude<
+  { [K in keyof T]: T[K] extends {} ? K : never }[keyof T],
+  undefined
+>
+type Compact<T> = Pick<T, CompactKeys<T>>
```

Comment:
You're right that this is indeed difficult to type, I cannot come up with a good solution. It's good that you tried to model the function's behaviour in the type system, however, `compactOptions` does in fact not change much about the shape of the argument, right?

So `<T>compactOptions(options: T): T` would be correct.

Also, the current types disregard cases like this:

```ts
interface Q {
  a: string
  b: number | undefined
}
```
From `Compact<Q>` we get `{ a: string }`, even though `Q` requires `b` to be present in all cases, even if undefined. Note that this is never an issue with `ApiClient.Options` because it does not have such properties.

Reply:
Good catch. It is, in fact, possible to construct a different type (say `Required`) where required (but possibly `undefined`) types are not erased by the `Compact` type. But this is potentially in conflict with how the runtime implementation works since types are erased. Without types in runtime, there is no way for the code to tell if the `undefined` value belongs to a property that is required or optional. Hence, we are forced to erase the property from the object, anyway - giving us an object of `Compact` type.

You're absolutely right about not changing the shape of the argument much though! Technically speaking, the object still conforms to `Partial<ApiClient.Options>` even if we erased all of its properties. There is a slight issue when we apply this to the example type `Q`, of course - because the property `b` is now (potentially) erased from the object. This isn't a strong issue if we are only using the value of `Q.b` - i.e. `undefined` -> `undefined` - but could be when we want to interact with keys of `Q` at runtime; typescript will still think that `Q` has property `b`, when it does not.

For the time being, I think I'd favor the simplicity of using type `T` even if it isn't a perfect representation, over an approach that conditionally converts only properties whose type includes `undefined` into optional types.

Reply:
I agree 100 %.

In fact, I also looked into `Required`, and you're right.

Given that the helper function won't be exported, it's probably not a big issue that using simply `T` is not perfect.